### DISCUSSION TITLE
Some ideas to simplify the code

### DISCUSSION THREAD

**Discussion by kqv1** [RomanArzumanyan/VALI#92] [Ideas]
I would like to suggest some random ideas:

1. Rename **EncodeSingleSurface**, **DecodeSingleSurface**, **DecodeSingleFrame** to **EncodeSurface**, **DecodeSurface**, **DecodeFrame**

2. Add overrides of **Surface.Make** to get its parameters from **Encoder** or **Decoder**. Surfaces usually exist in context of an existing **Encoder** or **Decoder**, except the **PixelFormat** which may be needed to change. 
The same idea can apply to **ColorspaceConversionContext** (to get its parameters from existing **Encoder** or **Decoder**).
Also, why do we use static **Surface.Make** function but not some constructor?
```python
surf_src = vali.Surface.Make(
    format=pyDec.Format(),
    width=pyDec.Width(),
    height=pyDec.Height(),
    gpu_id=0)
surf_dst = vali.Surface.Make(
    format=vali.PixelFormat.RGB,
    width=pyDec.Width(),
    height=pyDec.Height(),
    gpu_id=0)
cctx = ColorspaceConversionContext(pyDec.ColorSpace, pyDec.ColorRange)

############################################
# Can be simplified as
surf_src = vali.Surface(pyDec)
surf_dst = vali.Surface(pyDec, vali.PixelFormat.RGB)
cctx = vali.ColorspaceConversionContext(pyDec)
# or may be this
surf_src = pyDec.MakeSurface()
surf_dst = pyDec.MakeSurface.(vali.PixelFormat.RGB)
cctx = pyDec.MakeColorspaceConversionContext()
```

3. Add support for direct conversion from **NV12** to **RGB_PLANAR** as this format is frenquently used in AI. I assume that the underlying nVidia library doesn't support that conversion pair, but can we do it in **pybind11** code?

4. Many functions required **gpu_id** as a parameter. Should we set it as optional with the default **gpu_id=0**, or add module function **vali.SetDefaultGpu()**

If @RomanArzumanyan agree I can create a PR.

**Comment by RomanArzumanyan** [comment]
Hi @kqv1 
Thank you for your suggestions.

> Rename `EncodeSingleSurface`, `DecodeSingleSurface`, `DecodeSingleFrame` to `EncodeSurface`, `DecodeSurface`, `DecodeFrame`

Renaming is a very expensive because it breaks API. That's something we can only afford during major version bumps.

Also, there's a reason why `Single` exists in taxonomy. Previous decoder class `PyNvDecoder` was able to take encoded frame from  packet. That's useful for video analytics engines and surveillance applications as it allows to store compressed frames in RAM instead of raw frames in vRAM. I'd like to bring back this feature support to new `PyDecoder` class some day. And generally, there's no 1 to 1 ratio between number of packets and number of frames (single frame can be spread across multiple packets as well as single packet may contain multiple frames). In these circumstances, `DecodeSingleSurface` method was guaranteed to decode a single video frame precisely.

> Add overrides of `Surface.Make` to get its parameters from Encoder or Decoder.

IMO it's a good idea. I've been thinking about that for some time but didn't have a chance to implement it. If you like to submit a PR with this feature - please go ahead.

> Add support for direct conversion from NV12 to RGB_PLANAR as this format is frequently used in AI. I assume that the underlying nVidia library doesn't support that conversion pair, but can we do it in pybind11 code?

You are correct. There's no direct NV12 > RGB conversion supported in NPP. One possible solution is to do 2 separate conversions under the hood (NV12 > YUV420 + YUV420 > RGB) within the `PySurfaceConverter` class guts. I'd like to avoid writing the CUDA kernels within VALI because it will add significant support burden and will complicate the build system by introducing CUDA architectures selection code. There are other libraries such as CV-CUDA from Nvidia but it lacks color space and color range conversion parameters and may not be as accurate as NPP.

> Many functions required `gpu_id` as a parameter. Should we set it as optional with the default `gpu_id=0`, or add module function `vali.SetDefaultGpu()`

I'd like to avoid extending internal state with introduction of default GPU as a plague. This may introduce so many nasty invalid CUDA context bugs on so many levels. 

WRT the default `gpu_id`: one may actually use VALI on a system without GPU. `PyDecoder` and `PyFrameConverter` classes will be available, so basic functionality of "give me RGB frame for my AI code" is covered. So there's no GPU preference, hence I don't see much sense in introducing it by making `gpu_id=0` a default value. Also, explicit `gpu_id` is very simple to test.

**Reply by kqv1** [reply]
1-3-4/ I understand your view as the maintainer of the library, as the existing is already functioning perfect. My proposals are from the view of developers who use the library (I want my code to be as simple as possible). Maybe you can revisit item 3 and 4 in the future.

2/ I would suggest using constructor **Surface()** to create surface instead of static function **Surface.Make()**, but that would also break API

**Comment by RomanArzumanyan** [comment]
Hi @kqv1 

> Maybe you can revisit item 3 and 4 in the future

I'd like to get rid of NPP for the sake of CV-CUDA because later has async API which creates great performance improvements possibilities. However, CV-CUDA is in early stages of public development so I keep my fingers crossed.

> I would suggest using constructor Surface() to create surface instead of static function Surface.Make(), but that would also break API

Usual approach to this problem is to extend the existing API and then drop legacy code after couple major version bumps in order to give users some time to migrate.

So you can come up with 2 overloaded methods `Surface()` and `Surface.Make` which take as argument any class that supports `Width`, `Height` and `Format` methods. Just take `py::object` argument and check if it has these attributes (you may look up the DLPack code for that).

**Comment by kqv1** [comment]
2/ After consideration, I don't recommend passing `PyNvEncoder` / `PyDecoder` object. Now I recommend to add constructors for `Surface` class (same parameters as `Surface.Make`), as well as add `MakeSurface` methods to `PyNvEncoder` / `PyDecoder` classes. 

Users can create `Surface` by any:
```python
surf_src = vali.Surface.Make(pyDec.Format, pyDec.Width, pyDec.Height, 0) #current way, we can remove this one later
surf_src = vali.Surface(pyDec.Format, pyDec.Width, pyDec.Height, 0)
surf_src = pyDec.MakeSurface(vali.PixelFormat.RGB) #if the format is omitted then use pyDec.PixelFormat
```
4/In `PyDecoder.cpp` 4.2.5 version, [there is already default value](https://github.com/RomanArzumanyan/VALI/blob/606e8743b0f890e335acc66f4050cd70f274ebea/src/python_vali/src/PyDecoder.cpp#L298) for `gpu_id = 0`. I recommend to add `default gpu_id = 0` to all other `pybind11` declaration.

If @RomanArzumanyan agree I can create a PR.

**Comment by RomanArzumanyan** [comment]
Hi @kqv1 

```python
surf_src = pyDec.MakeSurface(vali.PixelFormat.RGB)
```

I'm not in favor of adding more responsibilities to decoder which is already much more complex then any other class.
Also, creating Surfaces of different format within the class that has it's format fixed doesn't make much sense to me.

```python
surf_src = vali.Surface(pyDec.Format, pyDec.Width, pyDec.Height, 0)
```

Refactoring `Surface.Make` to just `Surface` is ok. However, this can't be done in single update.
Users need some time to migrate to new API.

**Reply by kqv1** [reply]
> Also, creating Surfaces of different format within the class that has it's format fixed doesn't make much sense to me.

Creating Surface for other format (but the same resolution) could be convenient for computer vision applications, where user usually have to convert between different formats (mostly from NV12 to PLANAR anyway).

> I'm not in favor of adding more responsibilities to decoder which is already much more complex then any other class.

Understand your view. You may consider it later,


> Refactoring Surface.Make to just Surface is ok. However, this can't be done in single update.
> Users need some time to migrate to new API.

Sure, we should keep both ways for awhile


>default gpu_id = 0

Any comment to add this to the constructor of other classes?
